### PR TITLE
A40 - xDS Configuration Dump via Client Status Discovery Service in gRPC

### PR DESCRIPTION
**Pull request by lidizheng**
gRFC: https://github.com/lidizheng/proposal/blob/csds/A40-csds-support.md

### CODE REVIEW

File: A40-csds-support.md
Comment (srini100) [inline A40-csds-support.md:54]: extra '('

**Review by srini100 — APPROVED** [review]
LGTM but wait for @markdroth's review and approval.

File: A40-csds-support.md
Comment (menghanl) [inline A40-csds-support.md:109]: Add status (ACK/NACK) to the example.

File: A40-csds-support.md
Comment (menghanl) [inline A40-csds-support.md:114]: Break into multiple lines? This is pretty long, and hard to read.

File: A40-csds-support.md
Comment (menghanl) [inline A40-csds-support.md:126]: Would it better to link the PRs in each section?

File: A40-csds-support.md
Comment (menghanl) [inline A40-csds-support.md:238]: What's `(read more)`?

File: A40-csds-support.md
Comment (menghanl) [inline A40-csds-support.md:193]: Update this.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Good catch! It's https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#knowing-when-a-requested-resource-does-not-exist. Link updated.

File: A40-csds-support.md
Comment (sergiitk) [inline A40-csds-support.md:141]: (suggestion) Line numbers tend to change, consider replacing with a permalink.

**Review by sergiitk — APPROVED** [review]

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: You are right. I keep forgetting "servicer" is a Python-only term... Updated for all occurances.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:30]: Thanks for the reminding. Added the service method explanation part back.

(Just be told to keep gRFC short and concise in the other PR, so I removed many content that could be found without too much trouble.)

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:53]: Good idea. Link added.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:68]: In the most recent update https://github.com/envoyproxy/envoy/pull/14900, people decided to remove the per-xds-type status from CSDS's PerXdsConfig, and only use per-resource client status with type `ClientResourceStatus`.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Good idea. Updated to:

```
* If one resource won’t parse: don’t abort the parsing, record the error in the
  `error_state.details` field. The error string will be attached to other
  resources that are successfully recognized;
```

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Good catch. Updated.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Link updated.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:189]: Added a short note about gRPC SHOULD include the Node information:

```
Note that xDS config should include the `config.core.v3.Node` information
obtained from the bootstrap file in the
`envoy.service.status.v3.ClientConfig.node` field. Unlike other xDS configs,
this information is static and may require extra logic to integrate into the
CSDS service.
```

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Good idea! Updated!

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:241]: By returning `INVALID_ARGUMENT`, the RPC is terminated. The status only works as trailer headers, so it applies to all gRPC languages.

File: A40-csds-support.md
Comment (lidizheng) [inline A40-csds-support.md:259]: The requirement of Node information is updated.
CC @sergiitk @menghanl

Comment (sergiitk) [reply A40-csds-support.md:259]: Thanks for capturing this! Glad we've figured it out.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:6]: This isn't true yet.  We should add languages here as they actually get implemented.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:29]: Might be worth linking to https://github.com/envoyproxy/envoy/issues/13181.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:66]: The new `ClientResourceStatus` enum has already been accepted in the xDS protos, so we don't need to include it in this gRFC.  That's now part of the CSDS API, and we're just implementing that API.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:164]: I don't think we need to copy the proto message here.  The link above to the envoy protos is sufficient.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:191]: I don't really see a compelling reason to do this.  The reason that the node field is present in CSDS is that CSDS was originally designed to be served by control planes, not clients, and the control plane needed a way to indicate which response was for which client.  That's not relevant when you're talking to one individual client, so it's not clear to me that there's any benefit in the client populating this field.

File: A40-csds-support.md
Comment (markdroth) [reply A40-csds-support.md:259]: As I mentioned above, I don't see any compelling reason to bother with populating this field.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:268]: s/throw/return/

(This is not an exception, it's a status code.)

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:291]: This isn't an "if"; we've already done this.

**Review by markdroth — COMMENTED** [review]
Thanks for doing this!  My comments are mostly minor.

Please let me know if you have any questions.  Thanks!

File: A40-csds-support.md
Comment (sergiitk) [reply A40-csds-support.md:259]: @markdroth @lidizheng @menghanl and I discussed that in a chat. To summarize:
1) Parity with [`istioctl proxy-config bootstrap `](https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/#inspecting-bootstrap-configuration)
2) Some fields are generated after reading the bootstrap and sent to the control plane in requests: `user_agent_name`, `user_agent_version`, and, more importantly `clientFeatures`. Exposing them can be helpful for debugging. Note this is different from regular CSDS behavior to dump configs as is. The argument for treating the node info differently is that it's loaded from bootstrap and _sent_ to the control plane, as opposed to it's _received_ from the control plane

File: A40-csds-support.md
Comment (sergiitk) [reply A40-csds-support.md:191]: Added some context here: https://github.com/grpc/proposal/pull/223#discussion_r583764176

File: A40-csds-support.md
Comment (markdroth) [reply A40-csds-support.md:259]: If there's desire to expose the client's node information for debugging purposes, then that's a perfectly reasonable use-case, and we should design something to address it.  However, CSDS was not designed to solve that problem, and IMHO it would be a very clunky way to solve that problem.  If the mechanism we provide for accessing the bootstrap info is for a client to query CSDS, that means that a client that wants only the bootstrap info has no way to get it that does not also send the entire set of known xDS resources.  That's sending a potentially very large amount of completely unnecessary data.

If we want a way to expose the client's bootstrap info for debugging purposes, let's add a separate RPC service for that.  I don't think it should be part of this design.

Comment (sergiitk) [reply A40-csds-support.md:259]: I think I might be missing your point. Could you please clarify what data (if any) should be in [`ClientStatusResponse. ClientConfig.node`](https://github.com/envoyproxy/envoy/blob/98ac59243cf0c2960f267664da88c08f8b9641d1/api/envoy/service/status/v3/csds.proto#L126-L127)?

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Updated to `Core/C++/Java/Golang (1st release)`

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:29]: Good catch! I forgot the existence of this issue.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: This section is removed.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Removed.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Done.

(In Python, it is an exception)

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Removed the if statement.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:206]: > it would be a very clunky way to solve that problem

I agree the Node information will be sent along other xDS config dumps. And users might not need them all the time. But the same argument could apply to all xDS configs, like users just updated listeners and would like to see the listeners, and we are sending routes, clusters, endpoints in the CSDS response.

When the xDS implements get popularity soon, we might have users posting issues, and with the node information we could tell which gRPC language/release it is using, and (possibly) which control plane it is trying to talk to.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:138]: The Node information is helpful to debug gRPC and control plane simultaneously. In common cases, users connect to the CSDS endpoint  with ip:port, so the Node information can be new to them.

It also works if a user sets configuration in the control plane, and not observing the expected behavior. Then the user can use this to check what xDS config gRPC receives, and use the Node information to query the control plane's version of the xDS configs. This is how "diff" works in `istioctl` (https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/#retrieve-diffs-between-envoy-and-istiod).

**Review by lidizheng — COMMENTED** [review]
Thanks for the detailed review! PTALA

File: A40-csds-support.md
Comment (voidzcy) [inline A40-csds-support.md:79]: The handling is still a bit ambiguous. I guess we would only record the first error, so the parsing does not need to continue. But would attach the error message to all previously accepted resources that the current response is attempting to update.

Comment (lidizheng) [reply A40-csds-support.md:79]: Can you specify the cases that are ambiguous here? I guess the wording here is not accurate enough, if one resource (e.g., a cluster) has validation error, the parsing should continue; or one of its field has parsing error, e.g. a config in Any failed to deserialize, the parsing should continue. In other word, the parsing should continue when possible.

If you need a reference PR, please see https://github.com/grpc/grpc/pull/25329.

Comment (voidzcy) [reply A40-csds-support.md:79]: Looks C-core is [concatenating](https://github.com/markdroth/grpc/blob/0f9af3c045bd9a7d71348e3bd294565386c81c16/src/core/ext/xds/xds_api.cc#L1472) errors together if there are more than one error resources. I don't see much benefit of doing so, other than showing a giant error message while may still not showing all errors (e.g., multiple validation errors for a single resource). Anyway, I am fine with what's being implemented in C-core.

File: A40-csds-support.md
Comment (markdroth) [reply A40-csds-support.md:206]: @sergiitk 
> Could you please clarify what data (if any) should be in [`ClientStatusResponse. ClientConfig.node`](https://github.com/envoyproxy/envoy/blob/98ac59243cf0c2960f267664da88c08f8b9641d1/api/envoy/service/status/v3/csds.proto#L126-L127)?

What I am proposing is that we do not populate that field at all.

@lidizheng 
> But the same argument could apply to all xDS configs, like users just updated listeners and would like to see the listeners, and we are sending routes, clusters, endpoints in the CSDS response.

I don't see what that statement has to do with the argument I'm making here.  It's true that CSDS does not provide the ability to filter which set of configs are returned, but that's equally true whether CSDS is being served by the control plane or the client.  (And this would be trivial to add if someone needed it just by adding a resource type filter to the CSDS request proto.)

The argument I am making here is that the reason that the node field exists in the CSDS response is that CSDS was designed to be served by control planes, and the control plane will be returning data for multiple clients and needs a way to indicate which client each set of data is relevant to.  But when CSDS is sent on a client, all of the data is by definition for the same client, so there's no reason to populate this field.  In other words, the node field needs to be populated only when the CSDS server is a control plane; it is simply not relevant when the CSDS server is a client.

> When the xDS implements get popularity soon, we might have users posting issues, and with the node information we could tell which gRPC language/release it is using, and (possibly) which control plane it is trying to talk to.

IMHO, the fact that this could conceivably be useful to someone someday is not enough justification to add it.  If that was a good enough justification, why isn't it standard practice for every API to include the gRPC language and release?  Why doesn't channelz include this info?  Why doesn't every single RPC service designed by applications provide this info?

In case it's not clear, the answer is that if it's not what a given API is designed to do, it shouldn't be part of that API.  The microservices approach means that each API should provide only the data that is necessary for the problem that service is designed to solve.  If that wasn't the case, then all APIs would be huge, because everyone would throw everything they could think of into every one of them.

CSDS is designed to show the set of resources seen by the client, not the client's node information.  If we want something to expose the client's node information, that's IMHO a separate use-case, and we should design a separate service to solve that problem.

In any case, I don't think this is important enough to justify spending more time debating it, so I'll leave it up to you.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:6]: It is not actually implemented in any of these languages yet.  This field should be empty for now.

File: A40-csds-support.md
Comment (markdroth) [inline A40-csds-support.md:123]: The trailing `:` should now be a `.` instead.

File: A40-csds-support.md
Comment (markdroth) [reply A40-csds-support.md:138]: As mentioned elsewhere, I'm not arguing that this info might not be useful, just that CSDS is not the right way to expose it.  Exposing this info is a separate problem, and if we want to solve that, we should provide a separate RPC service for it.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md:79]: I think the important part is peeking into the ADS response to see what set of resources is being updated, then pasting the error details to their `error_state` field. I'm fine with one error, as long as they are logged in the right place in the CSDS response, so it's actionable for users.

File: A40-csds-support.md
Comment (menghanl) [inline A40-csds-support.md:158]: What to do if the underlying client is speaking v2?
Return an error? What error code to use? `FailedPrecondition`?

Comment (lidizheng) [reply A40-csds-support.md:158]: By underlying client, do you mean the gPRC application? I thought gRPC processes all xDS config into v3, so the dumped xDS config is agnostic to v2/v3. If the external CSDS client is only speaking v2, they won't be able to invoke the v3 method, because the fully-qualified-method-name includes the service name which has "v3" in it. So, the v2 external CSDS client will get UNIMPLEMENTED by default, in theory.

Comment (sergiitk) [reply A40-csds-support.md:158]: Here's what TD does for v3 (currently only v2 supported):

```
❯ csds-client -request_file request.yaml -api_version v3
2021/03/01 18:20:55 rpc error: code = Unimplemented desc = The GRPC target is not implemented on the server, host: trafficdirector.googleapis.com, method: /envoy.service.status.v3.ClientStatusDiscoveryService/StreamClientStatus.
```

Comment (lidizheng) [reply A40-csds-support.md:158]: Nice, so in practicce, the old client will also get UNIMPLEMENTED.

Comment (markdroth) [reply A40-csds-support.md:158]: Note that the CSDS RPC service name is different between v2 and v3, because the major version is part of the proto package name.  So I think all we need to do here is to implement only the v3 service, not the v2 service.  Then clients will get UNIMPLEMENTED if they try to use the v2 service, just like they would for any other unimplemented RPC service.

Comment (markdroth) [reply A40-csds-support.md:158]: Oh, sorry, Lidi already said exactly what I just said. :)

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Put to-be-done instead, delegated to the future self.

File: A40-csds-support.md
Comment (lidizheng) [reply A40-csds-support.md]: Done.

**Review by lidizheng — COMMENTED** [review]
@markdroth PTALAA. Is there any blocking issues?

File: A40-csds-support.md
Comment (menghanl) [reply A40-csds-support.md:158]: So the answer is, we support csds v3 only, csds v2 will get error with code unimplemented.
But the data can be either xds v2 or xds v3, right?

**Review by markdroth — APPROVED** [review]

File: A40-csds-support.md
Comment (menghanl) [reply A40-csds-support.md:158]: If the underlying data is xds v2, that means the `proto.Any`s in the xds responses are v2 resources.

And in csds, I would assume the dumped resources (`proto.Any`s) should be v3 type? That means we need to turn those v2 `proto.Any` resources from the xds responses into v3 `proto.Any` (unmarshal and then marshal again), and then put in the dumped messages, is that right?

(We alreay unmarshal the v2 resources into v3 types to read. So this basically means we need to then marshal the v3 type into `proto.Any` and keep those in cache as raw resources, instead of the `proto.Any`s from the xds responses. It's not too many extra steps. But just want to confirm this.)

Comment (lidizheng) [reply A40-csds-support.md:158]: I would vote for yes, let's convert the packed v2 resources into v3.

It would make analyzing the CSDS response simpler. Though, I can only think of `proto.Any` for ApiListener and HTTP filters. ApiListener will be converted to v3 for sure. We don't support v2 xDS HTTP filters at all, so they will be rejected in the first place. Do you have an example that where the v2 resources might slip through?

Comment (markdroth) [reply A40-csds-support.md:158]: I think it's fine to send v2 protos inside of the `Any` fields in v3 CSDS.  In xDS terms, the transport protocol version and the data model version do not have to be the same.  I don't think we need to do any conversion here.

Comment (menghanl) [reply A40-csds-support.md:158]: > Do you have an example that where the v2 resources might slip through

It's not one type of resource. All the resources in xds v2 are v2. E.g. `v2.Cluster` vs `v3.Cluster`, `v2.Endpoint` vs `v3.Endpoint`.

> In xDS terms, the transport protocol version and the data model version do not have to be the same.

That's right.
The difference would be on the csds client side. If we send v2 protos as `Any` in the csds responses, the csds client needs to depend on v2 protos (not that they will be used, but their types need to be registered) so the client can unmarshal those `Any`.

Comment (markdroth) [reply A40-csds-support.md:158]: I think it's fine for CSDS client to need v2 protos if the client is using v2.  That's independent of the fact that we're only supporting v3 of the CSDS RPC service itself.

This is no different than ADS -- it's possible for a client to use v3 of ADS but v2 of the resource protos, or vice versa.  In fact, in Envoy, those are two independent knobs.  In gRPC, we just simplified it and avoided one of the knobs by depending on the fact that we didn't happen to use any fields in v2 that were being removed in v3, so we could just unconditionally treat all of the resource protos as v3.

Comment (menghanl) [reply A40-csds-support.md:158]: This SGTM (and is actually easier to do :))

Let's update the gRFC to clarify this? This doesn't sound accurate:
> `This doc proposes
to not support xDS v2 for CSDS.`

Comment (lidizheng) [reply A40-csds-support.md]: Updated to
> For xDS CSDS v2 (the RPC service itself), this doc proposes to not support, so
> we only provide v3 CSDS service and CSDS client requesting for v2 service will
> get UNIMPLEMENTED error.
> 
> There are edge cases that v2 and v3 xDS resources are mixed together. This doc
> proposes to keep them as-is in the CSDS responses. So, the CSDS service can
> accurately reflect the xDS configs received from the control plane.

The section title also updated to `xDS v2/v3/v4`, since it's hard to describe our handling of v2 in the title.

**Review by menghanl — APPROVED** [review]

**Comment by sergiitk** [comment]
🥳

**Pull request merged**
